Bluetooth: GMAP: Add initial implementation of GMAP

Thalley opened the pull request:
Add initial implementation of GMAP.

This comes with the implementation of all GMAP roles, as well as BSIM and shell support. 

Samples are not yet implemented, but may be added as part of this PR or a subsequent PR.

Thalley commented:
The implementation of TMAP in https://github.com/zephyrproject-rtos/zephyr/pull/56686 and this should be made similar. TBD which changes to be done by either.

kartben commented on include/zephyr/bluetooth/audio/gmap.h:
```diff
@@ -0,0 +1,217 @@
+	 * Requires BT_GMAP_UGT_FEAT_SINK to be set as well
+	 */
+	BT_GMAP_UGT_FEAT_64KBPS_SINK = BIT(3),
+	/** @briefSupport for receiving multiple LC3 codec frames per block in an SDU
```
FWIW you may actually want to not even have `@brief` tags as it's implicit, just like e.g. you're not having it for the brief of the enum itself at L74

Thalley replied:
Does doxygen always implicitly consider the first paragraph (i.e. until the first empty line) as the `@brief`? I was under the impression that `@brief` should be used for any multi-line documentation. 

Either way, using `@brief` (vs not using it) is more consistent with existing codebase, don't you agree?

kartben replied:
Yep, consistency is good! :) I guess my point was actually about consistency, since you already have a mix of `@brief` and implicit briefs for the enum and structs in this file, so would be good to go for one or the other indeed.

> Does doxygen always implicitly consider the first paragraph (i.e. until the first empty line) as the `@brief`? 

The brief is the contents of the first line until the first dot (if there is one)

> I was under the impression that `@brief` should be used for any multi-line documentation.

Ya, it was until https://github.com/zephyrproject-rtos/zephyr/pull/60314 -- I made this change to help in exactly this 👇 kind of situation, where if people have already gone as far as using Doxygen style comment, then there is a 99.9% chance that the text is indeed a brief description, even if does not have the `@brief` tag

```
/** Gaming Role bitfield */
enum bt_gmap_role {
```

Thalley committed:
Bluetooth: GMAP: Add initial implementation of GMAP

Add initial implementation of Gaming Audio Profile (GMAP).

This is a top layer profile in the LE Audio stack,
designed for low latency audio ideal for gaming.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: GMAP: Add GMAP shell

Add the GMAP shell module with support for running the
various audio configurations in an easy way.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: GMAP: Add GMAP BSIM test

Add BSIM testing of GMAP, with all spec-specified audio
configuration combinations added and tested.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

carlescufi commented:
@MariuszSkamra @kruithofa @larsgk could you please review?

larsgk commented on tests/bsim/bluetooth/audio/src/gmap_ugg_test.c:
```diff
@@ -0,0 +1,1236 @@
+
+	err = bt_bap_unicast_client_register_cb(&unicast_client_cbs);
+	if (err != 0) {
+		FAIL("Failed to register CAP callbacks (err %d)\n", err);
```
unicast callbacks?

larsgk commented on tests/bsim/bluetooth/audio/src/gmap_ugg_test.c:
```diff
@@ -0,0 +1,1236 @@
+	}
+
+	if (src_ep_cnt != src_cnt) {
+		FAIL("Source  endpoint and stream count mismatch: %zu != %zu\n", src_ep_cnt,
```
extra space

larsgk commented on tests/bsim/bluetooth/audio/src/gmap_ugg_test.c:
```diff
@@ -0,0 +1,1236 @@
+		}
+	}
+
+	/* Set all endpoints from multiple connections in a single array, and verify that the known
+	 * endpoints matches the audio configuration
+	 */
+	for (size_t i = 0U; i < param->conn_cnt; i++) {
+		for (size_t j = 0U; j < param->snk_cnt[i]; j++) {
+			snk_cnt++;
+		}
+
+		for (size_t j = 0U; j < param->src_cnt[i]; j++) {
+			src_cnt++;
+		}
+	}
```
Comment and code doesn't seem to match for me for this block.

larsgk commented on subsys/bluetooth/audio/shell/gmap.c:
```diff
@@ -0,0 +1,564 @@
+#define UNICAST_SINK_SUPPORTED (CONFIG_BT_BAP_UNICAST_CLIENT_ASE_SNK_COUNT > 0)
+#define UNICAST_SRC_SUPPORTED  (CONFIG_BT_BAP_UNICAST_CLIENT_ASE_SRC_COUNT > 0)
+
+#define GMAP_AC_MAX_CONN   2U
```
Should this be a config or will it always be 2?

Thalley replied:
It will always be 2 based on the defined audio configurations from the GMAP spec

larsgk commented on subsys/bluetooth/audio/shell/gmap.c:
```diff
@@ -0,0 +1,564 @@
+static int cmd_gmap_ac_14(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct bap_broadcast_ac_param param = {
+		.name = "AC_13",
```
AC_14?

larsgk reviewed (changes requested):
Didn't try to run it (maybe we should - it's a fairly big PR).  Found a few things but otherwise looks ok to me

cvinayak commented:
@Thalley rebase required.
@larsgk please re-review

Thalley commented:
@fredrikdanebjer @kruithofa @cvinayak @MariuszSkamra @Casper-Bonde-Bose it would good to get this (experimental) feature into 3.5 if possible - If you can please review ASAP before the feature freeze tomorrow.

msmttchr commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -622,6 +622,8 @@ static void connected(struct bt_conn *conn, uint8_t err)
 		}
 	}
 
+	default_conn = bt_conn_ref(conn);
```
@Thalley, I think this line has been inserted by mistake as it does call twice `bt_conn_ref`. Is this intentional?

MariuszSkamra replied:
@msmttchr https://github.com/zephyrproject-rtos/zephyr/pull/66484
